# Week 4 Homework
#### This Directory is my record of using the tool [TERRACE](https://github.com/Shao-Group/TERRACE) on total paired-end RNAseq data of *Felis catus*


## [Installation](https://bioconda.github.io/recipes/terrace/README.html)
### Option 1 

You will need a conda-compatible package manager, like mamba

```bash

conda config --add channels conda-forge
conda config --add channels bioconda
conda install -c conda-forge mamba

```
Now Install TERRACE

```bash
mamba install terrace
```

confirm installation 

```bash
conda list terrace
```
This should return something like:
```bash
# Name                     Version          Build            Channel
terrace                    1.1.2            he153687_0       bioconda
```

## Make sure it is up-to-date
```bash
mamba update terrace
```

To create a new environment, run:
```bash
mamba create --name myenvname terrace
```



### Option 2
Installation with Docker container:
[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat)](http://bioconda.github.io/recipes/terrace/README.html)




## Downloading Data

Install sra-toolkit if not previously installed:
```bash
mamba install -c bioconda sra-tools
```
Install fastqc if not previously installed:

```bash
sudo apt install fastqc
```

Making directories to install data

```bash

mkdir Felis_catus_RNA
cd Felis_catus_RNA

```

RNA-seq of Felis catus: spleen
[SRR3218716](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR3218716&display=metadata)

```bash
vdb-config --interactive
fastq-dump -F --split-files SRR3218716
```



## [Usage](https://github.com/Shao-Group/TERRACE?tab=readme-ov-file#usage)



The usage of TERRACE is:
```bash
terrace -i <input.bam> -o <output.gtf> -fa <reference-genome.fa> --read_length <length-of-paired-end-reads> -r [reference_annotation.gtf] -fe [feature_file] [options]
```
The input.bam is the read alignment file generated by some RNA-seq aligner, (for example, STAR or HISAT2). Make sure that it is sorted; otherwise run samtools to sort it:
```bash
samtools sort input.bam > input.sort.bam
```














